### PR TITLE
[Docs] Revamping the InstructLab README

### PR DESCRIPTION
**Pull request by kelbrown20**
**Description:** Going through the InstructLab readme and consolidating, restructuring and rewording some content.

The preview might help with checking out the formatting: 
- https://github.com/kelbrown20/kelly-instructlab/tree/revamp-readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,107 +86,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 - Python 3.10 or Python 3.11
-- Approximately 60GB disk space (entire process)
 
-> **NOTE:** Python 3.12 is currently not supported, because some dependencies don't work on Python 3.12, yet.
-<!-- -->
-> **NOTE:** When installing the `ilab` CLI on macOS, you may have to run the `xcode-select --install` command, installing the required packages previously listed.
+   📋 Python 3.12 is currently not supported. Some InstructLab dependencies don't work on Python 3.12, yet.
 
-## ✅ Getting started
+- Approximately 60GB disk space (entire process)
```
Comment (nathan-weinberg) [inline README.md:98]: @cdoern is this still accurate do you think?

File: README.md [hunk 2]
```diff
@@ -86,107 +86,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 
-2. Create a new directory called `instructlab` to store the files the `ilab` CLI needs when running and `cd` into the directory by running the following command:
+> [!NOTE]
+> The following steps in this document use [Python venv](https://docs.python.org/3/library/venv.html) for virtual environments. However, if you use another tool such as [pyenv](https://github.com/pyenv/pyenv) or [Conda Miniforge](https://github.com/conda-forge/miniforge) for managing Python environments on your machine continue to use that tool instead. Otherwise, you may have issues with packages that are installed but not found in your virtual environment.
+
+### 🧰 Installing `ilab`
+
+1. Create a new directory called `instructlab` to store the files the `ilab` CLI needs when running and `cd` into the directory by running the following command:
```
Comment (nathan-weinberg) [inline README.md:115]: No longer true - we have dedicated directories for `ilab` at the system level now - @cdoern and @RobotSail can share details

File: README.md [hunk 3]
```diff
@@ -86,107 +86,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 
-3. There are a few ways you can locally install the `ilab` CLI. Select your preferred installation method from the following instructions. You can then install `ilab` and activate your `venv` environment.
+   ⚠️ The `python3` binary shown in the installation guides are the Python version that you installed in the above step. The command can also be `python3.11` or `python3.10` instead of `python3`. You can check Python's version by `python3 -V`.
 
-   > **NOTE**: ⏳ The `python3` binary shown in the following steps is the Python version that you installed in the above step. The command can also be `python3.11` or `python3.10` instead of `python3`. You can check Python's version by `python3 -V`.
-   >
-   > **NOTE**: ⏳ `pip install` may take some time, depending on your internet connection. In case installation fails with error ``unsupported instruction `vpdpbusd'``, prepend `CMAKE_ARGS="-DLLAMA_NATIVE=off"` to `pip install` command.
+   ⏳ `pip install` may take some time, depending on your internet connection. In case installation fails with error ``unsupported instruction `vpdpbusd'``, append `-C cmake.args="-DLLAMA_NATIVE=off"` to `pip install` command.
```
Comment (nathan-weinberg) [inline README.md:126]:
```suggestion
   ⏳ `pip install` may take some time, depending on your internet connection. In case the installation fails with error ``unsupported instruction `vpdpbusd'``, append `-C cmake.args="-DLLAMA_NATIVE=off"` to `pip install` command.
```

File: README.md [hunk 4]
```diff
@@ -86,107 +86,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 
-   See [the GPU acceleration documentation](./docs/gpu-acceleration.md) for how to
-   to enable hardware acceleration for interaction and training on AMD ROCm,
-   Apple Metal Performance Shaders (MPS), and Nvidia CUDA.
+   See [the GPU acceleration documentation](./docs/gpu-acceleration.md) for how to to enable hardware acceleration for interaction and training on AMD ROCm, Apple Metal Performance Shaders (MPS), and Nvidia CUDA.
```
Comment (nathan-weinberg) [inline README.md:128]: This doc may need to be updated as well: https://github.com/instructlab/instructlab/issues/2237

Comment (nathan-weinberg) [reply README.md:128]: Also, it seems the below sections get into this anyway?

File: README.md [hunk 5]
```diff
@@ -86,107 +86,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
-      source venv/bin/activate
-      pip cache remove llama_cpp_python
-      CMAKE_ARGS="-DLLAMA_CUDA=on -DLLAMA_NATIVE=off" pip install 'instructlab[cuda]'
-      pip install vllm@git+https://github.com/opendatahub-io/vllm@2024.08.01
+   ```shell
+   python3 -m venv --upgrade-deps venv
+   source venv/bin/activate
+   pip cache remove llama_cpp_python
+   CMAKE_ARGS="-DLLAMA_CUDA=on -DLLAMA_NATIVE=off" pip install 'instructlab[cuda]'
+   pip install vllm@git+https://github.com/opendatahub-io/vllm@2024.08.01
```
Comment (nathan-weinberg) [inline README.md:186]: I think this might be out of date - e.g. we have a `requirements-vllm-cuda.txt` file now - @cdoern @n1hility can either of you confirm?

File: README.md (hunk 2 above)
Comment (nathan-weinberg) [reply README.md:115]: There's also some detail about this further down in the doc

File: README.md [hunk 6]
```diff
@@ -361,45 +344,59 @@ After running `ilab config init` your directories will look like the following o
...
 
 ### 📥 Download the model
 
-- Run the `ilab model download` command.
+#### Downloading the default InstructLab models
 
-  ```shell
-  ilab model download
-  ```
+- Run the `ilab model download` command to download a compact pre-trained version of the [model](https://huggingface.co/instructlab/) (~4.4G) from HuggingFace.
```
Comment (nathan-weinberg) [inline README.md:359]: This also now downloads a Mistral GGUF - cc @cdoern for details

File: README.md [hunk 7]
```diff
@@ -442,11 +434,19 @@ After running `ilab config init` your directories will look like the following o
...
-   Ensure vllm is installed:
+- You can serve a non-default model (e.g. Mixtral-8x7B-Instruct-v0.1) with the following example command:
+
+   ```shell
+   ilab model serve --model-path ~/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf
+   ```
+
+- You can serve a non-default Safetensors model (e.g. granite-7b-lab). This requires a GPU.
+
+   Verify vllm is installed:
```
Comment (nathan-weinberg) [inline README.md:449]: Same comment w.r.t vllm as I made above

File: README.md [hunk 8]
```diff
@@ -510,57 +514,57 @@ Detailed contribution instructions can be found in the [taxonomy repository](htt
...
+   ```
 
 ### 🚀 Generate a synthetic dataset
 
 Before following these instructions, ensure the existing model you are adding skills or knowledge to is still running. Alternatively, `ilab data generate` can start a server for you if you provide a fully qualified model path via `--model`.
 
 1. To generate a synthetic dataset based on your newly added knowledge or skill set in [taxonomy](https://github.com/instructlab/taxonomy.git) repository, run the following command:
 
-   With GPU acceleration:
+   *With GPU acceleration (Default)*
```
Comment (nathan-weinberg) [inline README.md:541]: @cdoern @aakankshaduggal to confirm these SDG sections, I'm still learning about this

File: README.md [hunk 9]
```diff
@@ -617,130 +621,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
 There are many options for training the model with your synthetic data-enhanced dataset.
 
-> **Note:** **Every** `ilab` command needs to run from within your Python virtual environment.
+### ✋ Before you begin training
 
-#### ilab model training pipelines
+There are a few models you need to download before running the InstructLab end-to-end workflow locally.
 
-`ilab model train` has three pipelines: `simple`, `full`, and `accelerated`. The default is `accelerated`.
+- Download the `granite-7b-lab` model for training:
```
Comment (nathan-weinberg) [inline README.md:628]: Technically, users could train other models - I'd just change the wording to say something like "for example"

File: README.md [hunk 10]
```diff
@@ -617,130 +621,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
 
-```shell
-ilab model train
-```
+   ```shell
+   ilab model download --repository prometheus-eval/prometheus-8x7b-v2.0 --hf-token <your-huggingface-token> 
+   ```
 
-> **NOTE:** ⏳ This step can potentially take **several hours** to complete depending on your computing resources. Please stop `ilab model chat` and `ilab model serve` first to free resources.
+#### InstructLab model training pipelines
```
Comment (nathan-weinberg) [inline README.md:640]: I love this and would love to see one for SDG as well!

File: README.md [hunk 11]
```diff
@@ -617,130 +621,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+
+`ilab model train` has three pipelines: `simple`, `full`, and `accelerated`. The default is `accelerated`.
```
Comment (nathan-weinberg) [inline README.md:642]: I think it's `full`? @cdoern can you confirm?

File: README.md [hunk 12]
```diff
@@ -617,130 +621,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+   ├── special_tokens_map.json
+   ├── tokenizer.json
+   ├── tokenizer.model
+   └── tokenizer_config.json
+   ```
 
 #### Train the model locally with GPU acceleration
 
-Training has experimental support for GPU acceleration with Nvidia CUDA or AMD ROCm. Please see [the GPU acceleration documentation](./docs/gpu-acceleration.md) for more details. At present, hardware acceleration requires a data center GPU or high-end consumer GPU with at least 18 GB free memory.
+- Training has experimental support for GPU acceleration with Nvidia CUDA or AMD ROCm. Please see [the GPU acceleration documentation](./docs/gpu-acceleration.md) for more details. At present, hardware acceleration requires a data center GPU or high-end consumer GPU with at least 18 GB free memory.
```
Comment (nathan-weinberg) [inline README.md:737]: I wouldn't call it "experimental" at this point :smile: what say you @RobotSail?

File: README.md [hunk 13]
```diff
@@ -784,177 +801,132 @@ resources. You can download it via `ilab model download`.
...
-1. Stop the server you have running by entering `ctrl+c` keys in the terminal running the server.
+If you have a server running, stop the server you have running by entering `ctrl+c` keys in the terminal running the server.
 
-   > **IMPORTANT**:
-   - 🍎 This step is only implemented for macOS with M-series chips (for now).
+- 🍎 This step is only implemented for macOS with M-series chips (for now).
 
-   - Before serving the newly trained model you must convert it to work with
-   the `ilab` cli. The `ilab model convert` command converts the new model into quantized [GGUF](https://medium.com/@sandyeep70/ggml-to-gguf-a-leap-in-language-model-file-formats-cd5d3a6058f9) format which is required by the server to host the model in the `ilab model serve` command.
+Before serving the newly trained model you must convert it to work with
```
Comment (nathan-weinberg) [inline README.md:920]: I would rephrase this somehow - AFAIK conversion is something only Mac _needs_ to do, not necessarily a "feature" that is Mac-only - cc @cdoern

File: README.md (hunk 2 above)
Comment (kelbrown20) [reply README.md:115]: Yeah I didn't have to do this either, so Im good with removing it. I just wan't sure if it was a organization thing.

File: README.md (hunk 4 above)
Comment (kelbrown20) [reply README.md:128]: I think it uses the hardware by default when the config is created. So I may just remove this reference until that doc is updated. 
WDYT?

Comment (nathan-weinberg) [reply README.md:128]: Tbh I think it'd be better to address the issue as part of this PR - either updating that document or moving the content here and removing it entirely.

File: README.md (hunk 1 above)
Comment (cdoern) [reply README.md:98]: no, this is way off, probably requires 500GB

File: README.md (hunk 9 above)
Comment (kelbrown20) [reply README.md:628]: True, I only added this one because granite-7b-lab is the default model in the config. So the user should download it at some point. 

This part is more suppose to be like a Prereq

File: README.md (hunk 10 above)
Comment (kelbrown20) [reply README.md:640]: Im working on one in the SDG repo, but I need to understand them more lol

File: README.md (hunk 4 above)
Comment (cdoern) [reply README.md:121]: i honestly think we should unlink this doc, its super out of date and i have gotten dms regarding the info in it

File: README.md (hunk 5 above)
Comment (cdoern) [reply README.md:179]: yeah, but thats only if people clone the repo

File: README.md (hunk 8 above)
Comment (cdoern) [reply README.md:534]: I wouldn't mark this as the default, it runs on cpu unless you provide --gpus

File: README.md [hunk 14]
```diff
@@ -510,57 +507,57 @@ Detailed contribution instructions can be found in the [taxonomy repository](htt
...
 
 Before following these instructions, ensure the existing model you are adding skills or knowledge to is still running. Alternatively, `ilab data generate` can start a server for you if you provide a fully qualified model path via `--model`.
 
 1. To generate a synthetic dataset based on your newly added knowledge or skill set in [taxonomy](https://github.com/instructlab/taxonomy.git) repository, run the following command:
 
-   With GPU acceleration:
+   *With GPU acceleration (Default)*
 
    ```shell
    ilab data generate --pipeline full --gpus <NUM_OF_GPUS>
```
Comment (cdoern) [inline README.md:537]: the cfg has pipeline full in it, --gpus is only if they are on a gpu enabled system. I think our default should be CPU

File: README.md [hunk 15]
```diff
@@ -510,57 +507,57 @@ Detailed contribution instructions can be found in the [taxonomy repository](htt
...
+   ⏳ This can take from 15 minutes to 1+ hours to complete, depending on your computing resources.
 
    *Example output of `ilab data generate`*
 
    ```shell
    (venv) $ ilab data generate
    INFO 2024-07-30 19:57:44,093 numexpr.utils:161: NumExpr defaulting to 8 threads.
    INFO 2024-07-30 19:57:44,452 datasets:58: PyTorch version 2.3.1 available.
-   Generating synthetic data using 'simple' pipeline, '$HOME/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
+   Generating synthetic data using 'simple' pipeline, '$HOME/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.optimizied' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
```
Comment (cdoern) [inline README.md:560]:
```suggestion
   Generating synthetic data using 'full' pipeline, '$HOME/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
```

File: README.md (hunk 9 above)
Comment (cdoern) [reply README.md:621]: ehhhhhhhhh I dont think anything other than simple really works with non-granite models

File: README.md (hunk 11 above)
Comment (cdoern) [reply README.md:635]: yep its full

File: README.md (hunk 12 above)
Comment (cdoern) [reply README.md:730]: yeah, not experimental

File: README.md (hunk 13 above)
Comment (cdoern) [reply README.md:913]: yes, mac only. I wouldnt call it a feature. the full pipeline does this for you

**Review by cdoern — COMMENTED** [review]
just a few comments, thanks!

File: README.md (hunk 4 above)
Comment (nathan-weinberg) [reply README.md:121]: Not just unlink then, let's remove it outright. I just want to make sure no info from there is lost.

File: README.md (hunk 5 above)
Comment (nathan-weinberg) [reply README.md:179]: hmmm no way to pip install with VLLM? shouldn't it be part of the CUDA target?

File: README.md (hunk 9 above)
Comment (nathan-weinberg) [reply README.md:621]: Yeah let's just leave this then, for now

File: README.md (hunk 4 above)
Comment (kelbrown20) [reply README.md:121]: So do yall think it would be best to just delete the gpu-acceleration.md file?

Comment (nathan-weinberg) [reply README.md:121]: if all the relevant info is copied elsewhere, yes

File: README.md [hunk 16]
```diff
@@ -86,107 +85,104 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 
-1. When installing on Fedora Linux, install C++, Python 3.10 or 3.11, and other necessary tools by running the following command:
+- When installing on Fedora Linux, install C++, Python 3.10 or 3.11, and other necessary tools by running the following command:
 
    ```shell
    sudo dnf install gcc gcc-c++ make git python3.11 python3.11-devel
    ```
 
-   If you are running on macOS, this installation is not necessary and you can begin your process with the following step.
+   Some Python version management tools that build Python (instead of using a pre-built binary) may not by default build libraries implemented in C, and CPython when installing a Python version. This can result in the following error when running the `ilab data generate` command: `ModuleNotFoundError: No module named '_lzma'`. This can be resolved by building CPython during the Python installation with the `--enable-framework`. For example for `pyenv` on MacOS: `PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install 3.x`. You may need to recreate you virtual environment after reinstalling Python.
```
Comment (alinaryan) [inline README.md:108]: need to recreate you->need to recreate your

File: README.md (hunk 5 above)
Comment (kelbrown20) [reply README.md:179]: @nathan-weinberg is the PR your working on related to this?

File: README.md (hunk 1 above)
Comment (kelbrown20) [reply README.md]: Yeah that seems like alot 😕

File: README.md (hunk 5 above)
Comment (nathan-weinberg) [reply README.md:179]: i'm only changing the tag, my points in this thread remain

Comment (kelbrown20) [reply README.md:179]: gotcha! @cdoern do you have insight on this?

File: README.md (hunk 4 above)
Comment (kelbrown20) [reply README.md:121]: I just went over the gpu-acceleration.md and most of the install parts were copied over to the README. So that content is covered. Im just not sure about some of additional cuda/rocm arguments that im unfamiliar with. Might need to sync with someone just for clarifications,

**Review by nathan-weinberg — APPROVED** [review]

File: README.md [hunk 17]
```diff
@@ -517,57 +537,63 @@ Detailed contribution instructions can be found in the [taxonomy repository](htt
...
+   ⏳ This can take from 15 minutes to 1+ hours to complete, depending on your computing resources.
 
    *Example output of `ilab data generate`*
 
    ```shell
    (venv) $ ilab data generate
    INFO 2024-07-30 19:57:44,093 numexpr.utils:161: NumExpr defaulting to 8 threads.
    INFO 2024-07-30 19:57:44,452 datasets:58: PyTorch version 2.3.1 available.
-   Generating synthetic data using 'simple' pipeline, '$HOME/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
+   Generating synthetic data using 'full' pipeline, '$HOME/.cache/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
```
Comment (cdoern) [inline README.md:596]:
```suggestion
   Generating synthetic data using 'full' pipeline, 'mistral-7b-instruct-v0.2.Q4_K_M.gguf' model, './taxonomy' taxonomy, against http://localhost:8000/v1 server
```

File: README.md [hunk 18]
```diff
@@ -624,130 +650,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
 
-`ilab model train` has three pipelines: `simple`, `full`, and `accelerated`. The default is `accelerated`.
+- Download the `granite-7b-lab` model for training:
 
-1. `simple` uses an SFT Trainer on Linux and MLX on MacOS. This type of training takes roughly an hour and produces the lowest fidelity model but should indicate if your data is being picked up by the training process.
-2. `full` uses a custom training loop and data processing funcions for the granite family of models. This loop is optimizied for CPU and MPS functionality. Please use `--pipeline=full` in combination with `--device=cpu` (Linux) or `--device=mps` (MacOS). You can also use `--device=cpu` on a MacOS machine. However, MPS is optimized for better performance on these systems.
-3. `accelerated` uses the instructlab-training library which supports GPU accelerated and distributed training. The `full` loop and data processing functions are either pulled directly from or based off of the work in this library.
+   ```shell
+   ilab model download --repository intructlab/granite-7b-lab --hf-token <your-huggingface-token>
```
Comment (cdoern) [inline README.md:660]: hf token is not necessary for granite

Comment (cdoern) [inline README.md:660]:
```suggestion
   ilab model download --repository intructlab/granite-7b-lab
```

File: README.md [hunk 19]
```diff
@@ -624,130 +650,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
 
 #### Train the model locally on an M-series Mac or on Linux using the full pipeline
 
-To train the model locally on your M-Series Mac using our full pipeline and MPS or on your Linux laptop/deskop using CPU:
+- To train the model locally on your M-Series Mac using our full pipeline and MPS or on your Linux laptop/desktop using CPU:
 
-```shell
-ilab model train --pipeline full --device mps
-```
+   ```shell
```
Comment (cdoern) [inline README.md:703]: missing `--data-path`

File: README.md [hunk 20]
```diff
@@ -624,130 +650,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+   ilab model train --pipeline full --device mps
```
Comment (cdoern) [inline README.md:704]: same here

File: README.md [hunk 21]
```diff
@@ -624,130 +650,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+   ```
 
-```shell
-ilab model train --pipeline full --device cpu
-```
+   ```shell
+   ilab model train --pipeline full --device cpu
```
Comment (cdoern) [inline README.md:708]: same here, simple doesn't require a `--data-path` but full and accelerated do

File: README.md [hunk 22]
```diff
@@ -624,130 +650,143 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
 #### Train the model locally with GPU acceleration
 
-Training has experimental support for GPU acceleration with Nvidia CUDA or AMD ROCm. Please see [the GPU acceleration documentation](./docs/gpu-acceleration.md) for more details. At present, hardware acceleration requires a data center GPU or high-end consumer GPU with at least 18 GB free memory.
+- Training has support for GPU acceleration with Nvidia CUDA or AMD ROCm. Please see [the GPU acceleration documentation](./docs/gpu-acceleration.md) for more details. At present, hardware acceleration requires a data center GPU or high-end consumer GPU with at least 18 GB free memory.
 
-```shell
-ilab model train --pipeline accelerated --device cuda
-```
+   ```shell
+   ilab model train --pipeline accelerated --device cuda
```
Comment (cdoern) [inline README.md:769]: `--data-path` here as well

**Review by cdoern — COMMENTED** [review]
a few changes in training but looks good!!

File: README.md (hunk 18 above)
Comment (kelbrown20) [reply README.md:660]: wait really, it didn't work for me without it 

`intructlab/granite-7b-lab requires a HF Token to be set. Please use '--hf-token' or 'export HF_TOKEN' to download all necessary models`

Comment (kelbrown20) [reply README.md:660]: same here

Comment (cdoern) [reply README.md:660]: It looks like you are typing `instuctlab` so it probably is erroring out bc of that

Comment (kelbrown20) [reply README.md:660]: whoops, ty, Ill check that!

File: README.md [hunk 23]
```diff
@@ -86,107 +86,99 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
 
 ## 📋 Requirements
 
-- **🍎 Apple M1/M2/M3 Mac or 🐧 Linux system** (tested on Fedora).
-  We anticipate support for more operating systems in the future.
+- **🍎 Apple M1/M2/M3 Mac or 🐧 Linux system** (tested on Fedora). We anticipate support for more operating systems in the future.
+
+   📋  When installing the `ilab` CLI on macOS, you may have to run the `xcode-select --install` command, installing the required packages previously listed.
```
Comment (mairin) [inline README.md:91]: Are the required packages previously listed? I think they're below? (I'm a little confused on that phrase, "the required packages previously listed.")

Comment (mairin) [reply README.md:91]: Oh I see. It's moved up :) So you might want to rephrase, "the required packages listed below."

File: README.md [hunk 24]
```diff
@@ -86,107 +86,99 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
 
-1. When installing on Fedora Linux, install C++, Python 3.10 or 3.11, and other necessary tools by running the following command:
+- When installing on Fedora Linux, install C++, Python 3.10 or 3.11, and other necessary tools by running the following command:
 
    ```shell
    sudo dnf install gcc gcc-c++ make git python3.11 python3.11-devel
    ```
 
-   If you are running on macOS, this installation is not necessary and you can begin your process with the following step.
+   Some Python version management tools that build Python (instead of using a pre-built binary) may not by default build libraries implemented in C, and CPython when installing a Python version. This can result in the following error when running the `ilab data generate` command: `ModuleNotFoundError: No module named '_lzma'`. This can be resolved by building CPython during the Python installation with the `--enable-framework`. For example for `pyenv` on MacOS: `PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install 3.x`. You may need to recreate your virtual environment after reinstalling Python.
```
Comment (mairin) [inline README.md:108]: Would this work better as a call out since it might not hit all users?

File: README.md [hunk 25]
```diff
@@ -624,130 +650,155 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+   ```
+
+#### InstructLab model training pipelines
 
-> **NOTE:** ⏳ This step can potentially take **several hours** to complete depending on your computing resources. Please stop `ilab model chat` and `ilab model serve` first to free resources.
+`ilab model train` has three pipelines: `simple`, `full`, and `accelerated`. The default is `full`.
+
+1. `simple` uses an SFT Trainer on Linux and MLX on MacOS. This type of training takes roughly an hour and produces the lowest fidelity model but should indicate if your data is being picked up by the training process.
+2. `full` uses a custom training loop and data processing functions for the granite family of models. This loop is optimized for CPU and MPS functionality. Please use `--pipeline=full` in combination with `--device=cpu` (Linux) or `--device=mps` (MacOS). You can also use `--device=cpu` on a MacOS machine. However, MPS is optimized for better performance on these systems.
+3. `accelerated` uses the instructlab-training library which supports GPU accelerated and distributed training. The `full` loop and data processing functions are either pulled directly from or based off of the work in this library.
```
Comment (mairin) [inline README.md:675]: Great explanation, makes everything very clear!!

File: README.md [hunk 26]
```diff
@@ -624,130 +650,155 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
+#### Train the model locally
+
+- Train the model with your synthetic data with the `ilab model train` command:
+
+   ```shell
+   ilab model train
+   ```
+
+   ⏳ This step can potentially take **several hours** to complete depending on your computing resources. Please stop `ilab model chat` and `ilab model serve` first to free resources.
+
```
Comment (mairin) [inline README.md:696]: Might be helpful to recommend a spec. Moo deng was on a 64GB ram system and took around 24 hours, is that right @cdoern ? Can we throw out a stat like that so folks have a rough idea?

File: README.md [hunk 27]
```diff
@@ -624,130 +650,155 @@ Note that it is also possible to generate a synthetic dataset based on the entir
...
-├── config.json
-├── pytorch_model.bin
-├── pytorch_model.gguf
-├── pytorch_model-Q4_K_M.gguf
-├── special_tokens_map.json
-├── tokenizer.json
-├── tokenizer_config.json
-└── tokenizer.model
-```
+   ⏳ This process will take a while to complete. If you run for ~8 epochs it will take several hours.
```
Comment (mairin) [inline README.md:717]: How do I tune it down to say 1 epoch for fastness?

File: README.md [hunk 28]
```diff
@@ -465,44 +481,48 @@ After running `ilab config init` your directories will look like the following o
...
+   Then you can serve a Safetensors model:
+
    ```shell
    ilab model serve --model-path ~/.cache/instructlab/models/instructlab/granite-7b-lab
    ```
 
 ### 📣 Chat with the model (Optional)
 
-Because you're serving the model in one terminal window, you will have to create a new window and re-activate your Python virtual environment to run `ilab model chat` command:
+- Because you're serving the model in one terminal window, you will have to create a new window and re-activate your Python virtual environment to run `ilab model chat` command:
```
Comment (mairin) [inline README.md:492]: One tip that would be useful here is temperature settings... I don't think we have fine grained temperature settings yet but for a temperature of one you pass `-gm` to ilab chat and it will set the temperature to 0 which makes it the most accurate.

see https://github.com/instructlab/instructlab/blob/0f9ba00931b66d822f8fcecf29b6bf391f9f9412/src/instructlab/model/chat.py#L642

**Review by mairin — CHANGES_REQUESTED** [review]
I think the "required packages above" comment thing needs to be changed, the rest is flexible and could be fixed in another PR.

File: README.md (hunk 23 above)
Comment (kelbrown20) [reply README.md:91]: Nice catch!

File: README.md (hunk 26 above)
Comment (kelbrown20) [reply README.md:696]: Im thinking about making like a performance chart based on tests we've done in a separate doc.

File: README.md (hunk 27 above)
Comment (kelbrown20) [reply README.md:717]: I'll see if there is a flag but users can also adjust it in the config

File: README.md [hunk 29]
```diff
@@ -86,107 +86,99 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
...
+   source venv/bin/activate
+   pip install instructlab
+   ```
 
-      > **NOTE**: Make sure your system Python build is `Mach-O 64-bit executable arm64` by using `file -b $(command -v python)`,
-      > or if your system is setup with [pyenv](https://github.com/pyenv/pyenv) by using the `file -b $(pyenv which python)` command.
-      >
-      > **NOTE**: Also, ensure that both `python3 -c 'import platform; print(platform.machine())'` and `arch` return `arm64`. If `arch` returns `i386` or `python3` returns `x86_64`, it means your terminal or Python is running in Intel emulation mode, which isn't optimized for Apple Silicon. This is especially important if you have synced data from an Intel-based Mac to an Apple Silicon Mac (M1/M2/M3), as you may need to review and update your environment, especially tools like `Homebrew` or `Bash` that were previously installed under Intel.
+   *Additional Build Argument for Intel Macs*
+   If you have an Mac with an Intel CPU, you must add a prefix of
```
Comment (alinaryan) [inline README.md:147]: an->a

**Review by alinaryan — COMMENTED** [review]
just a nit, otherwise looks great!

**Review by mairin — APPROVED** [review]
